File: 2023/kotlin/src/Day02.kt
private data class RequiredGameCubes(val id: Int, val red: Int, val green: Int, val blue: Int)

private fun parseRequiredCubesPerGame(input: String) =
    input.lines().map { line ->
        RequiredGameCubes(
            id = line.ints().first(),
            red = Regex("(\\d+) red").findAll(line).map { it.groupValues[1].toInt() }.max(),
            green = Regex("(\\d+) green").findAll(line).map { it.groupValues[1].toInt() }.max(),
            blue = Regex("(\\d+) blue").findAll(line).map { it.groupValues[1].toInt() }.max()
        )
    }

private fun part1(input: String) =
    parseRequiredCubesPerGame(input).sumOf { game ->
        if (game.red <= 12 && game.green <= 13 && game.blue <= 14) game.id else 0
    }

private fun part2(input: String) =
    parseRequiredCubesPerGame(input).sumOf { game ->
        game.red * game.green * game.blue
    }

fun main() {
    val input = readInput("Day02")
    part1(input).println()
    part2(input).println()
}

File: 2023/swift/AdventOfCode2023/Sources/Day02.swift
struct Day02: AdventDay {
  var input: String

  struct RequiredGameCubes {
    let id: Int
    let red: Int
    let green: Int
    let blue: Int
  }

  func parseRequiredCubesPerGame(_ input: String) -> [RequiredGameCubes] {
    input.lines().map { line in
      return RequiredGameCubes(
        id: line.ints().first!,
        red: line.matches(of: /(\d+) red/).map { Int($0.output.1)! }.max() ?? 0,
        green: line.matches(of: /(\d+) green/).map { Int($0.output.1)! }.max() ?? 0,
        blue: line.matches(of: /(\d+) blue/).map { Int($0.output.1)! }.max() ?? 0
      )
    }
  }

  func part1() -> Any {
    parseRequiredCubesPerGame(input)
      .map { $0.red <= 12 && $0.green <= 13 && $0.blue <= 14 ? $0.id : 0 }
      .sum()
  }

  func part2() -> Any {
    parseRequiredCubesPerGame(input)
      .map { $0.red * $0.green * $0.blue }
      .sum()
  }
}
